### PR TITLE
Test test

### PR DESCRIPTION
**Pull request by wlemkows**
- [ ] We still need to set up our runners with the correct HW to run this workflow successfully.

### CODE REVIEW

File: cmake/match.py [hunk 1]
```diff
@@ -22,10 +22,17 @@
...
 
-if len(match_lines) != len(input_lines):
-    sys.exit(f"Line count doesn't match (is: {len(input_lines)}, expected: {len(match_lines)})")
+if len(match_lines) < len(input_lines):
+    sys.exit(f"Match length < input length (input: {len(input_lines)}, match: {len(match_lines)})")
 
+input_idx = 0
+optional_line = False
 for i, match_line in enumerate(match_lines):
+    opt = "${OPT}"
```
Comment (pbalcer) [inline cmake/match.py:31]: hm, `{{OPT}}` to be consistent with how regex is specified?

File: cmake/match.py [hunk 2]
```diff
@@ -22,10 +22,17 @@
     input_lines = input.readlines()
     match_lines = match.readlines()
 
-if len(match_lines) != len(input_lines):
-    sys.exit(f"Line count doesn't match (is: {len(input_lines)}, expected: {len(match_lines)})")
+if len(match_lines) < len(input_lines):
+    sys.exit(f"Match length < input length (input: {len(input_lines)}, match: {len(match_lines)})")
 
+input_idx = 0
+optional_line = False
```
Comment (lukaszstolarczuk) [inline cmake/match.py:29]: shouldn't this setting be done within `for` loop?

// I know we don't have a case where we mix opt with non-opt lines, but just to be sure 😉

File: cmake/match.py [hunk 3]
```diff
@@ -35,9 +42,26 @@
...
+    input_line = input_lines[input_idx].strip()
     if not re.fullmatch(pattern, input_line):
-        print(f"Line {i+1} does not match".format(i+1))
-        print("is:       " + input_line)
-        print("expected: " + match_line.strip())
-        sys.exit(1)
+        if optional_line is True:
+            continue
+        else:
+            print(f"Line {i+1} does not match".format(i+1))
```
Comment (lukaszstolarczuk) [inline cmake/match.py:59]: hmm maybe I'm mistaken, but it seems you're mixing F-String with `str.format()` syntax 😉 either one or another...?

File: test/conformance/CMakeLists.txt [hunk 4]
```diff
@@ -8,9 +8,18 @@ set(UR_CONFORMANCE_TEST_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 function(add_test_adapter name adapter)
     set(TEST_TARGET_NAME test-${name})
     set(TEST_NAME ${name}-${adapter})
+
     add_test(NAME ${TEST_NAME}
-        COMMAND ${TEST_TARGET_NAME}
-        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+        COMMAND ${CMAKE_COMMAND}
+        -D TEST_FILE=${Python3_EXECUTABLE}
```
Comment (lukaszstolarczuk) [inline test/conformance/CMakeLists.txt:14]: Is this solution like a temporary one? Or should we maintain this list of "failing" tests forever, even if all of them are green?

Perhaps we should mention somewhere (at least in a comment here) what's the purpose of this extra matching here...

**Review by lukaszstolarczuk — COMMENTED** [review]
Can you please add some info in commit message? Explain a little why it's needed 😉

File: test/conformance/CMakeLists.txt (hunk 4 above)
Comment (pbalcer) [reply test/conformance/CMakeLists.txt:14]: Ideally all adapters would have 100% pass rate, but that's not the case at the moment unfortunately. And it's going to take some time for everything to get fixed. So yes, this is permanent until such time that we fix everything and agree to never break CTS again, which might be a while.

Definitely should be documented somewhere.

File: test/conformance/context/context_adapter_cuda.match [hunk 5]
```diff
@@ -0,0 +1 @@
+urContextCreateWithNativeHandleTest.Success/NVIDIA_CUDA_BACKEND___NVIDIA_GeForce_RTX_3060_
```
Comment (pbalcer) [inline test/conformance/context/context_adapter_cuda.match:1]: {{.*}} for device names? Same everywhere else.

**Comment by lukaszstolarczuk** [comment]
@wlemkows, please use my latest workflow commit from: https://github.com/lukaszstolarczuk/unified-runtime/commits/adapters-hw-ci

File: test/conformance/cts_exe.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+from argparse import ArgumentParser
+import subprocess  # nosec B404
+
+if __name__ == '__main__':
+
+    parser = ArgumentParser()
+    parser.add_argument("--test_command", help="Ctest test case")
+
+    args = parser.parse_args()
+    result = subprocess.Popen([args.test_command, '--gtest_brief=1'], stdout = subprocess.PIPE, text = True)  # nosec B603
```
Comment (pbalcer) [inline test/conformance/cts_exe.py:21]: `subprocess.run` ?

Comment (wlemkows) [reply test/conformance/cts_exe.py:21]: Popen is a better version of `run`, I use it here on purpose because as a result, I get the TextIOWrapper class on which I can perform additional operations that are not possible on the str object in the case of  `run`.

File: cmake/match.py (hunk 2 above)
Comment (wlemkows) [reply cmake/match.py]: Done.

File: cmake/match.py (hunk 3 above)
Comment (wlemkows) [reply cmake/match.py]: Done.

File: test/conformance/CMakeLists.txt (hunk 4 above)
Comment (wlemkows) [reply test/conformance/CMakeLists.txt:14]: Done.

**Comment by wlemkows** [comment]
> @wlemkows, please use my latest workflow commit from: https://github.com/lukaszstolarczuk/unified-runtime/commits/adapters-hw-ci


Done.

**Comment by wlemkows** [comment]
> Can you please add some info in commit message? Explain a little why it's needed 😉

Done.

File: .github/workflows/cmake.yml [hunk 7]
```diff
@@ -263,4 +313,4 @@ jobs:
         -DUR_FORMAT_CPP_STYLE=ON
         -DUMF_ENABLE_POOL_TRACKING=ON
     - name: Build
-      run: cmake --build ${{github.workspace}}/build -j $(nproc)
+      run: cmake --build ${{github.workspace}}/build -j $(nproc)
```
Comment (lukaszstolarczuk) [inline .github/workflows/cmake.yml:316]: suddenly EOF - pls revert this change 😉

File: cmake/match.py (hunk 1 above)
Comment (lukaszstolarczuk) [reply cmake/match.py]: Not quite. I think Piotr suggested without `$` - to be in line with what you have a few lines below, in the code:
`# split into parts at {{ }}`

File: cmake/match.py [hunk 8]
```diff
@@ -22,10 +22,18 @@
...
     match_lines = match.readlines()
 
-if len(match_lines) != len(input_lines):
-    sys.exit(f"Line count doesn't match (is: {len(input_lines)}, expected: {len(match_lines)})")
+if len(match_lines) < len(input_lines):
+    sys.exit(f"Match length < input length (input: {len(input_lines)}, match: {len(match_lines)})")
 
+input_idx = 0
 for i, match_line in enumerate(match_lines):
+    opt = "${{OPT}}"
```
Comment (lukaszstolarczuk) [inline cmake/match.py:30]: heh, this opt can be moved outside of the loop, actually

File: test/conformance/README.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Conformance tests
+
```
Comment (lukaszstolarczuk) [inline test/conformance/README.md:2]: maybe it's an obvious statement, but I'd add info that "when you fix tests you have to update match files" and perhaps mention why we have empty match files...?

File: test/conformance/CMakeLists.txt (hunk 4 above)
Comment (lukaszstolarczuk) [reply test/conformance/CMakeLists.txt:14]: Thanks!

File: test/conformance/cts_exe.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+#! /usr/bin/env python3
```
Comment (pbalcer) [inline test/conformance/cts_exe.py:1]: For now it's fine, but eventually, I'd like to see this script somehow merged with #810. They fulfill a similar role...

**Review by pbalcer — APPROVED** [review]

File: test/conformance/cts_exe.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+if __name__ == '__main__':
+
+    parser = ArgumentParser()
+    parser.add_argument("--test_command", help="Ctest test case")
+
+    args = parser.parse_args()
+    result = subprocess.Popen([args.test_command, '--gtest_brief=1'], stdout = subprocess.PIPE, text = True)  # nosec B603
+
+    while True:
```
Comment (PatKamin) [inline test/conformance/cts_exe.py:23]: This could be simplified to sth like:
```
for line in result.stdout.readlines():
...
```

File: test/conformance/cts_exe.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        line = result.stdout.readline()
+        if not line: break
+        if "[  FAILED  ]" in line:
+            test_case = line.split(" ")[5]
+            if (test_case[-1] == ","):
+                test_case = test_case[:len(test_case)-1]
+                print(test_case)
+    exit(0)
```
Comment (PatKamin) [inline test/conformance/cts_exe.py:31]: Possibly this exit is redundant, the script should exit with zero code at this point.

File: cmake/match.py [hunk 13]
```diff
@@ -35,9 +43,26 @@
...
+            print("expected: " + match_line.strip())
+            sys.exit(1)
+
+    input_line = input_lines[input_idx].strip()
     if not re.fullmatch(pattern, input_line):
-        print(f"Line {i+1} does not match".format(i+1))
-        print("is:       " + input_line)
-        print("expected: " + match_line.strip())
-        sys.exit(1)
+        if optional_line is True:
```
Comment (PatKamin) [inline cmake/match.py:57]: As I understand, this makes optional lines hard to check for - they can contain literally anything and the test can pass. Can we make opt lines required given a certain test scenario or setup?

Comment (pbalcer) [reply cmake/match.py:57]: That would require us to somehow make the lines "conditional", so something like `{{skip if compiler==clang}}`. While that might be useful, it's imho out of scope for this work. And no, optional lines can't be anything - this `continue` here will only move forward the match file, not the input file, so if the line doesn't match either the optional line or the next one, it will fail.

File: test/conformance/cts_exe.py (hunk 10 above)
Comment (wlemkows) [reply test/conformance/cts_exe.py:1]: Sure, we can plan such changes in the future.

**Review by pbalcer — APPROVED** [review]

File: test/conformance/cts_exe.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    args = parser.parse_args()
+    result = subprocess.Popen([args.test_command, '--gtest_brief=1'], stdout = subprocess.PIPE, text = True)  # nosec B603
+
+    pat = re.compile(r'\[( )*FAILED( )*\]')
+    for line in result.stdout.readlines():
+        if pat.search(line):
+            test_case = line.split(" ")[5]
+            if (test_case[-1] == ","):
+                test_case = test_case[:len(test_case)-1]
```
Comment (PatKamin) [inline test/conformance/cts_exe.py:30]: Try simpler `test_case.rstrip(',')`

File: test/conformance/cts_exe.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+            print(test_case)
+
+    output = result.communicate()
```
Comment (PatKamin) [inline test/conformance/cts_exe.py:33]: Missing output print?

**Review by lukaszstolarczuk — APPROVED** [review]

File: test/conformance/cts_exe.py (hunk 14 above)
Comment (wlemkows) [reply test/conformance/cts_exe.py]: Done.

File: test/conformance/cts_exe.py (hunk 15 above)
Comment (wlemkows) [reply test/conformance/cts_exe.py]: At this point I do not intend to use this variable, I just want to call the communicate function. Removed.

**Review by PatKamin — APPROVED** [review]
